In [1]:
#Importing Libararies
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
#Loading data
customer_data = pd.read_csv("Dataset/Churn_Modelling.csv")

In [4]:
#Display the first 5 rows of the Dataframe
customer_data.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

*Credit score :* credit score is a numerical representation of a person's creditworthiness, or how likely they are to repay borrowed money. It is used by banks and financial institutions to assess the risk of lending to a borrower.

*tenure :* tenure refers to the length of time for which a loan, deposit, or financial product is held or agreed upon. It represents the duration of the agreement between the borrower and the lender or the investor and the bank.

Exited is the Target variable

In [5]:
#Getting list of columns
columns = customer_data.columns.values.tolist()
print(columns)

['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']


In [6]:
#Removing unnecessary columns (rown numbers, ID column, surname)
#axis=1 for columns
#axis =0 for rows
dataset = customer_data.drop(["RowNumber","CustomerId","Surname"], axis=1)

In [7]:
dataset.head(5)

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0          619    France  Female   42       2       0.00              1   
1          608     Spain  Female   41       1   83807.86              1   
2          502    France  Female   42       8  159660.80              3   
3          699    France  Female   39       1       0.00              2   
4          850     Spain  Female   43       2  125510.82              1   

   HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0          1               1        101348.88       1  
1          0               1        112542.58       0  
2          1               0        113931.57       1  
3          0               0         93826.63       0  
4          1               1         79084.10       0

In [8]:
#Converting catergorical columns to numerical columns
#First we remove the catergorical columns ("Geneder" and "Geography")

dataset = dataset.drop(["Geography","Gender"],axis=1)

In [9]:
Geography = pd.get_dummies(customer_data.Geography).iloc[:,1:]
Gender = pd.get_dummies(customer_data.Gender).iloc[:,1:]

In [10]:
dataset = pd.concat([dataset,Geography,Gender],axis=1)

In [11]:
dataset.head(5)

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42       2       0.00              1          1   
1          608   41       1   83807.86              1          0   
2          502   42       8  159660.80              3          1   
3          699   39       1       0.00              2          0   
4          850   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  Germany  Spain  Male  
0               1        101348.88       1        0      0     0  
1               1        112542.58       0        0      1     0  
2               0        113931.57       1        0      0     0  
3               0         93826.63       0        0      0     0  
4               1         79084.10       0        0      1     0

In [17]:
#Checking for missing values
dataset.isnull().sum()

CreditScore        0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
Germany            0
Spain              0
Male               0
dtype: int64

In [25]:
#Creating Target variable
X = dataset.drop(["Exited"],axis=1)
y = dataset["Exited"]

In [26]:
#Checking for multicollinearity
#ideally vif greater than 5 is considered as high
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif = pd.DataFrame()
vif ['VIF'] = [variance_inflation_factor(X.values, i) for i in range (X.shape[1])]
vif["variables"] = X.columns

In [27]:
vif

VIF        variables
0   21.236445      CreditScore
1   12.334128              Age
2    3.872755           Tenure
3    3.182267          Balance
4    7.826417    NumOfProducts
5    3.289605        HasCrCard
6    2.075966   IsActiveMember
7    3.887186  EstimatedSalary
8    1.787170          Germany
9    1.486247            Spain
10   2.168988             Male

In [28]:
#correlation between 2 variables
correlation_matrix = dataset.corr()
print(correlation_matrix['Age'].sort_values(ascending=False))


correlation_matrix = dataset.corr()
print(correlation_matrix['CreditScore'].sort_values(ascending=False))


#vif values shows there is multicollineraity . i.e correlation between multiple variables

Age                1.000000
Exited             0.285323
IsActiveMember     0.085472
Germany            0.046897
Balance            0.028308
Spain             -0.001685
CreditScore       -0.003965
EstimatedSalary   -0.007201
Tenure            -0.009997
HasCrCard         -0.011721
Male              -0.027544
NumOfProducts     -0.030680
Name: Age, dtype: float64
CreditScore        1.000000
IsActiveMember     0.025651
NumOfProducts      0.012238
Balance            0.006268
Germany            0.005538
Spain              0.004780
Tenure             0.000842
EstimatedSalary   -0.001384
Male              -0.002857
Age               -0.003965
HasCrCard         -0.005458
Exited            -0.027094
Name: CreditScore, dtype: float64


In [29]:
#Data splitting into Train and Test dataset
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test =  train_test_split(X,y, test_size=0.2, random_state=0)

**MACHINE LEARNING ALGORITHMS**

*Random Forest*

In [ ]:
#Machine Learing Algorithm Training (Random Forest)
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators=200,random_state=0)
classifier.fit(X_train,y_train)
predictions = classifier.predict(X_test)

In [31]:
#Machine Learing Algorithm Training
#Accuracy Metrices
from sklearn.metrics import classification_report, accuracy_score

print(classification_report(y_test,predictions))
print(accuracy_score(y_test,predictions))

              precision    recall  f1-score   support

           0       0.89      0.95      0.92      1595
           1       0.73      0.51      0.60       405

    accuracy                           0.86      2000
   macro avg       0.81      0.73      0.76      2000
weighted avg       0.85      0.86      0.85      2000

0.8635


The Random Claasifier Model has an accuracy of 86.35%.


*XGBoost*

In [43]:
#Machine Learing Algorithm Training (XGboost)
from xgboost import XGBClassifier
model = XGBClassifier()
model.fit(X_train,y_train)
predictions = model.predict(X_test)
print(classification_report(y_test,predictions))
print(accuracy_score(y_test,predictions))

ModuleNotFoundError: No module named 'xgboost'